dao-voting/src/main.sw

```diff
@@ -0,0 +1,272 @@
+    ///
+    /// # Panics
+    ///
+    /// The function will panic when:
+    /// - The constructor is called more than once
+    /// - The voting period is not greater than 0
+    /// - The approval percentage is not greater than 0
+    #[storage(read, write)]
+    fn constructor(gov_token: ContractId) -> bool {
+        require(storage.state == 0, Error::CannotReinitialize);
+
+        storage.gov_token = gov_token;
+        storage.proposal_count = 0;
+        storage.state = 1;
+
+        true
+    }
+
+    /// Add proposal to be voted on
+    ///
+    /// # Panics
+    ///
+    /// The function will panic when:
+    /// - The constructor has not been called to initialize
+    /// - The voting period is 0
+    /// - The approval percentage is 0
+    #[storage(read, write)]
+    fn add_proposal(voting_period: u64, approval_percentage: u64, proposal_data: CallData) -> bool {
+        require(storage.state == 1, Error::NotInitialized);
+        require(voting_period > 0, Error::PeriodCannotBeZero);
+        require(approval_percentage > 0, Error::ApprovalPercentageCannotBeZero);
+
+        let proposal = Proposal {
+            yes_votes: 0,
+            no_votes: 0,
+            approval_percentage: approval_percentage,
+            call_data: proposal_data,
+            end_height: height() + voting_period,
+        };
+        storage.proposals.insert(storage.proposal_count, proposal);
+        storage.proposal_count = storage.proposal_count + 1;
+        true
+    }
+
+    /// Update the user balance to indicate they have deposited governance tokens.
+    /// A successful deposit unlocks voting functionality.
+    ///
+    /// # Panics
+    ///
+    /// The function will panic when:
+    /// - The constructor has not been called to initialize
+    /// - The user deposits an asset that is not the specified governance token.
+    /// - The user does not deposit and assets
+    #[storage(read, write)]
+    fn deposit() -> bool {
+        require(storage.state == 1, Error::NotInitialized);
+        require(storage.gov_token == msg_asset_id(), Error::NotGovernanceToken);
+        require(msg_amount() > 0, Error::NoAssetsSent);
+
+        let result: Result<Identity, AuthError> = msg_sender();
+        let sender: Identity = result.unwrap();
+
+        let prev_balance = storage.balances.get(sender);
+        let new_balance = prev_balance + msg_amount();
+        storage.balances.insert(sender, new_balance);
+
+        true
+    }
+
+    /// Update the user balance to indicate they have withdrawn governance tokens
+    ///
+    /// # Panics
+    ///
+    /// This functions will panic when:
+    /// - The constructor has not been called to initalize
+    /// - The user tries to withdraw more than their balance
+    #[storage(read, write)]
+    fn withdraw(amount: u64) -> bool {
+        require(storage.state == 1, Error::NotInitialized);
+
+        let result: Result<Identity, AuthError> = msg_sender();
```

Comment:
While not critical, it might be nice to have a utils file with a function that returns the sender as this is repeated in a lot of places